flink-runtime/src/main/java/org/apache/flink/runtime/operators/coordination/RecreateOnResetOperatorCoordinator.java

```diff
@@ -74,7 +74,8 @@ public void start() throws Exception {
     @Override
     public void close() throws Exception {
         closed = true;
-        coordinator.closeAsync(closingTimeoutMs);
+        // Wait for coordinator close before destructing any user class loader.
+        coordinator.closeAsync(closingTimeoutMs).get();
```

Comment:
That is the only commit which I copied from https://github.com/apache/flink/pull/14239 without actually reproducing the bug. Previously @becketqin explained to me why it can avoid `ClassNotFoundException` and why it is OK to block here.

@becketqin could you help answer the questions here?

Reply:
I do not remember the exact original issue I saw there. It could be that the `SplitEnumerator` is still doing something (e.g. partition discovery) before it is closed.

I am not sure if we need to worry about this here. For two reasons:

1. In Flink the closure of all the components has always been synchronous, so closing the `OperatorCoordinator` here synchronously seems fine. This call may only block if the `SplitEnumerator` itself blocks on closure. Although this might be something we cannot not avoid if there is user code, it is very likely the synchronous closure is still non-blocking. For example, `CheckpointCoordinator.shutdown()` may block on the closure of master hooks.

2. It is the JM's responsibility to ensure the components are closed normally to avoid potential resource leak. Theoretically speaking, we can make all the component closure asynchronous so the main thread of JM is totally non-blocking, but chances are it still has to wait until all the components are closed normally before it can move to the next step. So to some extent, blocking on `closure` is a correctness requirement.

Reply:
@becketqin Thanks for the explanation!

It looks like we don't have clear idea on the root cause of `ClassNotFoundException`. And there is potential risk (probably unlikely) with making the call synchronous when user code has blocking operation.

Regarding the 2nd point above, it looks like there could be tradeoff between resource leak and blocking JobManager's scheduler thread. I suppose it is better to understand the resource leak before making JobManager's thread blocking.

And from https://issues.apache.org/jira/browse/FLINK-20114, it looks like the ClassNotFoundException is emitted in the WARNING without disrupting user's job. It seems OK to still emit this ClassNotFoundException in the log until we resolve the concerns described above.

@StephanEwen @becketqin To place on the safe side, I have removed this change from this PR.